### PR TITLE
docs: Add base Docs for Code Styles and Code Patterns

### PR DESCRIPTION
**Pull request by CodeNameNinja**
Add Documentation for Code Styles and Code Patterns.

+ Added Documentation on Code Styles.
+ Added Documentation on Code Patterns
+ Added Guidelines for Clean Code and Refactoring.
+ Added Guideline for the File System

### CODE REVIEW

File: content/engineering/processes/file-structure-guide.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+| View      | User Interface                                                                         | Views/\<Context>/\<Name>       | Views/Auth/Login.xaml                                          |
+| Model     | Business Objects that encapsulate data and behavior of application domain              | Models/\<Context>/\<Name>      | Models/Employee/EmployeeMail.cs Models/Employee/EmployeeObj.cs |
+| ViewModel | Link between Model and View OR It Retrieves data from Model and exposes it to the View | ViewModels/\<Context>/\<Name>  | ViewModels/Auth/LoginViewModel.cs                              |
+| Interface | a syntactical contract that an entity should conform to                                | Interfaces/\<Name>             | Interfaces/User                                                |
+| Control   | An Resuable UI Element                                                                 | UI/Controls/\<Context>/\<Name> | UI/Controls/Button/FlatButton.xaml                             |
```
Comment (waletoye) [inline content/engineering/processes/file-structure-guide.md:19]: @CodeNameNinja  "a"

File: content/engineering/processes/development-processes.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+- Frontend applications are automatically deployed to staging when merging to `main`
+- Backend applications are deployed to a dedicated staging environment when being merged to `main`
+- Backend applications are deployed to production when invoking the `release` workflow
+- We use _strict_ [semantic versioning](https://semver.org/) on _all_ packages. Please learn this and implement it
+  strictly
+- Below version `1.0.0`, the API is considered unstable
+
+## Commit Messages
+
+- use the imperative, present tense: "change" not "changed" nor "changes"
```
Comment (waletoye) [inline content/engineering/processes/development-processes.md:23]: "Use"

File: content/engineering/processes/development-processes.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+- No capitals and no fullstops (.) at the end
+- When a PR is ready for merge we trigger a squash and merge which will squash all your commits into one, when doing that follow the pattern of `type/scope/ticket-number: description (#PR-number)`. i.e. `style/onboarding/PROJECT-123: change button color to red (#321)`.
+
+## Pull Requests
+
+- Unit & e2e tests and snapshot updates should be performed prior to pushing your branch to the repository. This limits the the number of unnecessary CI runs and is a time and cost saving measure. More commits, fewer pushes.
+- PR title must follow the pattern of `type/scope/ticket-number: description`. i.e. `style/onboarding/PROJECT-123: change button color to red`.
```
Comment (waletoye) [inline content/engineering/processes/development-processes.md:30]: @CodeNameNinja  this is different from current PR template that we use.

please check it out

File: content/engineering/processes/recommendations.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+- Always use and store date-time as **UTC time**
+- **Use unix timestamps** as they are the lowest common denominator, easy to generate, easy to validate, are UTC by default, the smallest data over the wire and in storage
+- **Store or retrieve user timezone/locale date independently**, when required for user locale based operations on the server (ie: timed notifications)
+- Always store as **locale rather than offset**, as offsets can be inconsistent due to day light saving etc.
+- [unixtimestamp.com](https://www.unixtimestamp.com/)
+
+## Use BEM
+
+We try to use BEM (Block Element Modifier) for everything we can, not just class names, but test IDs as well. Here is a good introduction: [BEM 101](https://css-tricks.com/bem-101/).
+
```
Comment (yskorge) [inline content/engineering/processes/recommendations.md:20]: Doesnt apply to this projecct

File: content/engineering/processes/pull-request-template.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+- [ ] Implementation - The functionality meets the acceptance criteria of the ticket
+- [ ] Executed - The code has been executed and appears functional
+- [ ] Environments configured - Any configuration updates have been added to the relevant environments
+- [ ] Code documentation - Newly added code is sufficiently documented
+- [ ] Readme - Necessary readme updated added
+
+## Notes
+
+Is there any other prudent information you would want to add?
+```
```
Comment (yskorge) [inline content/engineering/processes/pull-request-template.md:36]: If we want to go with this, please update on our handheld repo

File: content/engineering/processes/ADRs/code-styles.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+- We need a structed set of rules to define how we go about code styling/formating
+
+## Decision
+
+- Using Sonar Cube in our CI process to enforce code styles in the HandHeld Codebase
+- Using ReSharper/Rider for local linting and fomating
+
+## Consequences
+
+- Formatting and linting will be enforced in the CI process
```
Comment (yskorge) [inline content/engineering/processes/ADRs/code-styles.md:24]:
```suggestion
- Formatting and linting will be enforced in the CI process using Sonar linting for detepcting coding issues in real-time and get clear guidance on how to fix them.
```

File: content/engineering/processes/ci-cd.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
+---
+title: "CI CD"
```
Comment (yskorge) [inline content/engineering/processes/ci-cd.md:2]: We already have a CI/CD process section, please add info there or take the existing information and add it to this section :-)

File: content/engineering/processes/development-processes.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+title: "Development Processes"
+date: 2021-08-31T18:10:57+02:00
+draft: false
+description: "Development processes and coding standards"
+weight: 1
+---
+
+## Git
+
+- We follow a gitflow like strategy. All changes are made in branches based off of `main`
```
Comment (yskorge) [inline content/engineering/processes/development-processes.md:11]: we dont have a main, we use master

File: content/engineering/processes/development-processes.md (hunk 3 above)
Comment (yskorge) [reply content/engineering/processes/development-processes.md:30]: we should discsuss this tomorrow, I dont like the above pattern, but Paul would like us to stick to 1 format company wide

File: content/engineering/processes/development-processes.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+- We use [the angular standard](https://gist.github.com/brianclements/841ea7bffdb01346392c) for our defined types
+
+## Coding standards
+
+- We use [eslint](https://www.npmjs.com/package/eslint) for JS/TS linting
+- We use [airbnb-standard](https://www.npmjs.com/package/eslint-config-airbnb-standard) for all projects
+- We use [prettier](https://www.npmjs.com/package/prettier) for automatic code formatting
+- We use [stylelint](https://www.npmjs.com/package/stylelint) for CSS-like linting
+- These are immutable style guides and avoid [bikeshedding](https://thedecisionlab.com/biases/bikeshedding)
+- Code needs to be written to support unit testing, ie. Dependency Injection
```
Comment (yskorge) [inline content/engineering/processes/development-processes.md:58]: none of this is applicable for this repo

File: content/engineering/processes/definition-of-ready-done.md [hunk 10]
```diff
@@ -1,46 +0,0 @@
----
-title: "Definition of Ready/Done"
-date: 2021-08-31T18:11:15+02:00
-draft: false
-description: "Definition of Ready and Done for each stage in the workflow"
-weight: 2
```
Comment (yskorge) [inline content/engineering/processes/definition-of-ready-done.md]: @CodeNameNinja Why was this removed? I like this section :D Just the miro board was linking to the incorrect board.

File: content/engineering/guides/naming-guideline.md [hunk 11]
```diff
@@ -0,0 +1,115 @@
+---
+title: "Naming Convention Guidelines"
+date: 2021-08-31T18:11:15+02:00
+draft: false
+description: "A guideline to naming conventions, including file names, variables, functions, classes, and more"
+weight: 1
+---
```
Comment (yskorge) [inline content/engineering/guides/naming-guideline.md:7]: Great 🦾

File: content/engineering/processes/pull-request-template.md (hunk 5 above)
Comment (CodeNameNinja) [reply content/engineering/processes/pull-request-template.md:36]: I think we should stick with our current template

File: content/engineering/processes/ADRs/code-styles.md [hunk 12]
```diff
@@ -0,0 +1,26 @@
+---
+title: "Code Styles"
+date: 2021-09-01T17:05:10+02:00
+draft: false
+description: "This is an ADR for our finding regarding Code Styles"
+weight: 1
+---
```
[the remark this answers is not in the document]
Comment (CodeNameNinja) [reply content/engineering/processes/ADRs/code-styles.md:7]: This will be added in a seperate PR

**Review by yskorge — APPROVED** [review]

**Review by waletoye — APPROVED** [review]
